certbot-dns-rfc2136/certbot_dns_rfc2136/dns_rfc2136.py

```diff
@@ -0,0 +1,215 @@
+"""DNS Authenticator using RFC 2136 Dynamic Updates."""
+import logging
+
+import dns.flags
+import dns.message
+import dns.name
+import dns.query
+import dns.rdataclass
+import dns.rdatatype
+import dns.tsig
+import dns.tsigkeyring
+import dns.update
+import zope.interface
+
+from certbot import errors
+from certbot import interfaces
+from certbot.plugins import dns_common
+
+logger = logging.getLogger(__name__)
+
+
+@zope.interface.implementer(interfaces.IAuthenticator)
+@zope.interface.provider(interfaces.IPluginFactory)
+class Authenticator(dns_common.DNSAuthenticator):
+    """DNS Authenticator using RFC 2136 Dynamic Updates
+
+    This Authenticator uses RFC 2136 Dynamic Updates to fulfull a dns-01 challenge.
+    """
+
+    ALGORITHMS = {
+      'HMAC-MD5': dns.tsig.HMAC_MD5,
+      'HMAC-SHA1': dns.tsig.HMAC_SHA1,
+      'HMAC-SHA224': dns.tsig.HMAC_SHA224,
+      'HMAC-SHA256': dns.tsig.HMAC_SHA256,
+      'HMAC-SHA384': dns.tsig.HMAC_SHA384,
+      'HMAC-SHA512': dns.tsig.HMAC_SHA512
+    }
+
+    description = 'Obtain certs using a DNS TXT record (if you are using BIND for DNS).'
+    ttl = 120
+
+    def __init__(self, *args, **kwargs):
+        super(Authenticator, self).__init__(*args, **kwargs)
+        self.credentials = None
+
+    @classmethod
+    def add_parser_arguments(cls, add):  # pylint: disable=arguments-differ
+        super(Authenticator, cls).add_parser_arguments(add, default_propagation_seconds=60)
```

Comment:
I was testing with two public DNS servers; the master that accepts the dynamic updates and another that just slaves the whole zone from the master. Both are listed as authoritative.

Even with using notifies to minimise the delay between the two servers when there was an update it sometimes still took the slave a while before performing the actual zone transfer upon receipt of a notify. I think BIND puts pending zone transfers into a queue and performs them serially so if there are a lot of zones and/or updates it can sometimes take a bit longer to propagate.

It did work when I still had a propagation delay of 10 seconds but that was just pure luck as the ACME server was choosing the master DNS server for queries. 60 seconds seemed to always be sufficient for everything to settle but it's probably something that's fairly site-specific.

For testing, there's not too much that needs to be set up in BIND. You need to generate a TSIG key:
```
dnssec-keygen -a HMAC-SHA512 -b 512 -n HOST keyname.
```
Then in your `named.conf` you need something like:
```
key "keyname." {
        algorithm hmac-sha512;
        secret "<the Base64-encoded string from the .key file generated above>";
};

zone "example.com." IN {
        type master;
        file "named.example.com";
        // You can either do this which allows the key full access to the zone:
        // allow-update { key "keyname."; };
        // Or you can grant the key access to only a very specific part of the zone:
        update-policy { grant keyname. name _acme-challenge.host.example.com. txt; };
};
```
If you need any help in setting up BIND let me know. You can also use PowerDNS which also supports the same update mechanism but I've not tried that.